book.md

```diff
@@ -177,7 +177,146 @@ vulnerabilities, along with their mitigations, presented in a rough
+program and parsing the output of the victim program are usually done in an
+external program that drives the exploit. This could be, for example, a C
+program communicating with the victim over the Internet. In this case, the
+primitives are abstracted into C functions, which prepare and send packets to
+the victim, and parse the victim's responses. These functions can be easily
+chained together, interleaved with arbitrary computations, all written in C, to
+form the exploit.
+
+It's interesting to note that while the read/write primitives consist of
+carefully constructed inputs to the victim program, the attacker can view these
+inputs as *instructions* to the victim program). This duality is
+explored in [@Dullien2020].
+
+In the case of a non-interactive attack, all computation happens within the
+victim program. The duality of input data and code is even more obvious in this
+case, as the malicious input to the victim can be viewed as the exploit code.
+There are cases for which the input is obviously interpreted as code by the
+victim application as well, as in the case of a JavaScript program given as
+input to a JavaScript engine. In this case, the read/write primitives would
+be written as JavaScript functions, which when called have the unintended
+side-effect of accessing arbitrary memory that a JavaScript program is not
+supposed to have access to.  The primitives can be chained together with
+arbitrary computations, also expressed in JavaScript.
+
+There are, however, cases, where the correspondence between data and code isn't
+as obvious. For example, in [@Beer2021], the malicious input consists of a PDF
+file, masquerading as a GIF. Due to an integer overflow bug in the PDF decoder,
+the malicious input leads to an unbounded buffer access.  In the case of
+JavaScript engine exploitation, the attacker would normally be able to use
+JavaScript operations and perform arbitrary computations, making exploitation
+more straightforward. In this case, there are no scripting capabilities
+officially supported. The attackers, however, take advantage of the compression
+format intricacies to implement a small computer architecture, in thousands of
+simple commands to the decoder.  In this way, they effectively _introduce_
+scripting capabilities and are able to express their exploit as a program to
+this architecture.
+
+We have described read/write primitives, and have discussed how an attacker
+might also perform arbitrary computations, either in an external program
+in the case of interactive attacks, or by using scripting capabilities
+(whether originally supported or introduced by the attacker) in
+non-interactive attacks. Assuming an attacker has gained these capabilities,
+how can they use them to achieve their goals?
+
+The ultimate goal of an attacker may vary: it may be, among other things,
+getting access to a system, leaking sensitive information or bringing down a
```

Comment:
could leaking sensitive information or bringing down a service be considered subtypes of getting access to a system, I think might have the wrong idea here?

Reply:
No I think they're helpful. In hindsight, I was over-analyzing it. I tried for a while to come up with some alternative wordings and clarify my thoughts, but in the end the text that you have already covers it all. LGTM.